gnucash/import-export/aqb/gnc-flicker-gui.c

```diff
@@ -0,0 +1,408 @@
+/*
+ * gnc-flicker-gui.c --
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of
+ * the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, contact:
+ *
+ * Free Software Foundation           Voice:  +1-617-542-5942
+ * 51 Franklin Street, Fifth Floor    Fax:    +1-617-542-2652
+ * Boston, MA  02110-1301,  USA       gnu@gnu.org
+ */
+
+/**
+ * @internal
+ * @file gnc-flicker-gui.c
+ * @brief GUI callbacks for Flicker and ChipTAN(optisch)
+ * @author Copyright (C) 2020 Christian Wehling <christian.wehling@web.de>
+ */
+
+#include <cairo.h>
+#include <gtk/gtk.h>
+#include <glib/gi18n.h>
+
+#include "gnc-flicker-gui.h"
+
+/* structured data for the flicker variables */
+struct _flickerdraw {
+	const char *challenge;
+	guint maxWert;
+    guint margin;           /* Abstand zwischen Balken */
+    guint barwidth;         /* Balkenbreite */
+    guint barheight;        /* Balkenhöhe */
+    guint x_barpos;			/* x-Wert für die Position des Balken */
+    guint y_barpos;			/* y-Wert für die Position des Balken */
+    guint x_drawpos;		/* Erste Zeichenposition */
+    guint y_drawpos;		/* Erste Zeichenposition */
+    guint height;			/* Höhe der Zeichenfläche */
+    guint width;			/* Breite der Zeichenfläche */
+    guint delay;            /* Wartezeit zwischen Frames in Millisekunden */
+    guint halfbyteid;
+    guint clock;
+    guint interval;
+    gboolean change_interval;
+    gint win_height;		/* Höhe des Dialogs */
+    gint win_width;			/* Breite des Dialogs */
+} flickerdraw;
+
+static GncFlickerGui *flickergui = NULL;
+_Bool bitarray[100][5];
```

Comment:
Which data type is preferred, g_boolean or uint8_t? Both work in the program.

Reply:
In the bank specification, the length of the challenge is specified with up to 255 characters. Will that exceed the storage requirements?
In the case of a transmission, I've got by with around 80 characters so far.

Reply:
I guess you didn't understand "recant". `_Bool` *is* the right choice if you really do need the clock bit, otherwise you could use either `unsigned char` or `uint8_t` (they're interchangeable) to store two hex digits, which would simplify the loading function a bit as you wouldn't need the `bits` array.

If the spec says 255 then `bitarray` should be 255; that's only 157 bytes, nothing to worry about.